### PR TITLE
Event organizer can now cancel the event + cancellation mail

### PR DESCRIPTION
**Pull request by rpg600**

### CODE REVIEW

File: src/AppBundle/Event/EventCancelledEvent.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    public function __construct(Adherent $author, Event $event, Committee $committee = null)
+    {
+        parent::__construct($committee);
+
+        $this->author = $author;
+        $this->event = $event;
+    }
+
+    public function getAuthor()
```
Comment (hhamon) [inline src/AppBundle/Event/EventCancelledEvent.php]: Missing return type hint.

File: src/AppBundle/Event/EventMessageNotifier.php [hunk 2]
```diff
@@ -38,16 +42,33 @@ public function onEventCreated(EventCreatedEvent $event)
...
+            $event->getEvent(),
+            $event->getAuthor()
+        ));
+    }
+
+    public function onEventCancelled(EventCancelledEvent $event)
+    {
+        $committee = $event->getCommittee();
+
+        if (!$committee) {
```
Comment (hhamon) [inline src/AppBundle/Event/EventMessageNotifier.php]: `if (!$committee = $event->getCommittee()) {`

File: src/AppBundle/Repository/AdherentRepository.php [hunk 3]
```diff
@@ -217,4 +219,26 @@ private function createManagedByQueryBuilder(Adherent $referent)
 
         return $qb;
     }
+
+    /**
+     * Finds the list of adherent registered for a given event.
```
Comment (hhamon) [inline src/AppBundle/Repository/AdherentRepository.php]: s/adherent/adherents/

File: tests/AppBundle/Controller/EventControllerTest.php [hunk 4]
```diff
@@ -157,6 +158,17 @@ public function testAnonymousUserCannotEditEvent($path)
 
     /**
      * @group functionnal
+     * @dataProvider provideCancelledProtectedPages
+     */
+    public function testRegisteredAdherentUserCannotAccessPagesOfCancelledEvent($path)
+    {
+        $this->authenticateAsAdherent($this->client, 'benjyd@aol.com', 'HipHipHip');
+        $this->client->request('GET', $path);
```
Comment (hhamon) [inline tests/AppBundle/Controller/EventControllerTest.php]: Use the constant from the Request class for the GET method.

File: tests/AppBundle/Mailjet/Message/AbstractEventMessageTest.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+use AppBundle\Entity\Event;
+use AppBundle\Entity\PostAddress;
+
+abstract class AbstractEventMessageTest extends \PHPUnit_Framework_TestCase
+{
+    protected function createEventMock(string $name, string $beginAt, string $street, string $cityCode): Event
+    {
+        $address = PostAddress::createFrenchAddress($street, $cityCode)->getInlineFormattedAddress('fr_FR');
+
+        $event = $this->getMockBuilder(Event::class)->disableOriginalConstructor()->getMock();
```
Comment (hhamon) [inline tests/AppBundle/Mailjet/Message/AbstractEventMessageTest.php]: `$event = $this->createMock(Event::class);`

Comment (hhamon) [inline tests/AppBundle/Mailjet/Message/AbstractEventMessageTest.php]: This new syntax is the shortcut for `$this->getMockBuilder(Event::class)->disableOriginalConstructor()->getMock();`

File: src/AppBundle/Controller/EventController.php [hunk 6]
```diff
@@ -55,6 +56,7 @@ public function exportIcalAction(Event $event): Response
     /**
      * @Route("/inscription", name="app_committee_attend_event")
      * @Method("GET|POST")
+     * @Security("event.isActive()")
```
Comment (rpg600) [inline src/AppBundle/Controller/EventController.php]: It's weird, you want to throw a 404 instead of a 403 ?

Comment (hhamon) [inline src/AppBundle/Controller/EventController.php]: Sounds more logic to me.

Comment (tgalopin) [inline src/AppBundle/Controller/EventController.php]: Indeed, a 404 is well suited IMO

File: src/AppBundle/Repository/EventRepository.php [hunk 7]
```diff
@@ -54,6 +54,19 @@ public function findOneByUuid(string $uuid): ?Event
         return $this->findOneBy(['uuid' => $uuid]);
     }
 
+    public function findOneActiveByUuid(string $uuid): ?Event
+    {
+        $query = $this->createQueryBuilder('e')
+            ->select('e')
```
Comment (hhamon) [inline src/AppBundle/Repository/EventRepository.php]: unneeded to specify `select()`

File: app/migrations/Version20170321155653.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        $this->addSql('ALTER TABLE events ADD status VARCHAR(20) NOT NULL');
+    }
+
+    public function postUp(Schema $schema)
+    {
+        /** @var Connection $connection */
+        $connection = $this->connection;
+
+        $connection->executeUpdate('UPDATE events e SET e.status = :status', ['status' => Event::STATUS_SCHEDULED], ['status' => \PDO::PARAM_STR]);
```
Comment (tgalopin) [inline app/migrations/Version20170321155653.php]: Is there a reason to put this here and not in the migration itself?

File: src/AppBundle/Controller/EventController.php [hunk 9]
```diff
@@ -139,6 +143,37 @@ public function editAction(Request $request, Event $event): Response
...
+     * @Method("GET|POST")
+     * @Entity("event", expr="repository.findOneActiveByUuid(uuid)")
+     * @Security("is_granted('HOST_EVENT', event)")
+     */
+    public function cancelAction(Request $request, Event $event): Response
+    {
+        $command = EventCommand::createFromEvent($event);
+
+        $form = $this->createForm(FormType::class);
+
```
Comment (tgalopin) [inline src/AppBundle/Controller/EventController.php]: Can you remove this empty line?

File: src/AppBundle/Entity/Event.php [hunk 10]
```diff
@@ -250,4 +263,29 @@ public function isFull(): bool
...
     }
+
+    public function getStatus(): string
+    {
+        return $this->status;
+    }
+
+    public function setStatus(string $status)
+    {
+        $this->status = $status;
```
Comment (tgalopin) [inline src/AppBundle/Entity/Event.php]: Can you throw an exception if the status is not valid?

File: app/migrations/Version20170321155653.php (hunk 8 above)
Comment (rpg600) [inline app/migrations/Version20170321155653.php]: Do you mean moving this line in the up method ?

**Pull request merged**